cirq-core/cirq/ops/conditional_operation.py

```diff
@@ -0,0 +1,143 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import (
+    AbstractSet,
+    Any,
+    cast,
+    Dict,
+    FrozenSet,
+    Optional,
+    Sequence,
+    TYPE_CHECKING,
+    Tuple,
+    Union,
+)
+
+from cirq import protocols, value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class ConditionalOperation(raw_types.Operation):
+    """Augments existing operations to be conditionally executed."""
+
+    def __init__(
+        self,
+        sub_operation: 'cirq.Operation',
+        conditions: Sequence[Union[str, 'cirq.MeasurementKey']],
+    ):
+        keys = tuple(value.MeasurementKey(k) if isinstance(k, str) else k for k in conditions)
+        self._control_keys = cast(Tuple[value.MeasurementKey, ...], keys)
+        self._sub_operation = sub_operation
```

Comment:
We *could* but I'd suggest not, as there's not really a reason to. TaggedOperation *has* to, since tags don't propagate from the suboperation by design. I think for ConditionalOperation it's cleaner just to leave it as-is.

Reply:
I disagree about there not being a reason to flatten. Nested `ConditonalOperation`s cause bloat in string-output methods (e.g. `CO(keys=a, sub_op=CO(keys=b, sub_op=op))` vs `CO(keys=[a, b], sub_op=op)`) and are more expensive to operate on than flat `ConditionalOperation`s (e.g. when using the recursive `unconditionally` method).

Reply:
Done, but note that ControlledOperation is the one that does this in the constructor. TaggedOperation does this in the with_tags function. So TaggedOperation(TaggedOperation(op), 'a'), 'b') would have only 'b' as a tag. I used the ControlledOperation approach. We should follow up and make this consistent.